### ISSUE TITLE
simplify And/Eq/Le

### ISSUE THREAD

**Issue by albertz** [sympy/sympy#18642]
```
i = Symbol("i", integer=True)
n = Symbol("n", integer=True)
expr = And(Eq(i, 2 * n), Le(i, 2*n -1))
```
I would expect that `expr.simplify()` yields `False`, but it does not.

**Comment by harshavardhan-99** [comment]
I would like to work on this issue, can you please elaborate on the issue a bit more?

**Comment by oscargus** [comment]
I think that one can add a simplification for `Eq` and `And` expressions where one replace one variable of the `Eq` with the other in all the other expressions. If that evaluates to `False`, then one can say that the expression is `False`, if it evaluates to  `True`, it should be enough to keep the `Eq` expression(s), as in #18645, and in case it returns a generic expression, one should not do anything, so basically a heuristic to see if the expression is even possible to be True at some stage. Not sure how useful it will be in general though.

I think that there may be a bit of this in the `Piecewise` simplifier, but probably only if one of the numbers are 

(Another option would be to enable `as_set()` to work with multiple variables.)

**Comment by oscarbenjamin** [comment]
It would be good to have a more general way of handling systems of at least linear equations and inequalities using e.g. linear programming (#21854).

The polynomial case is handled by cylindrical algebraic decomposition https://en.wikipedia.org/wiki/Cylindrical_algebraic_decomposition

**Comment by oscargus** [comment]
Indeed! I've been looking into minimizing polyhedrons, but didn't really find a good solution. The general case is quite tricky, especially doing it symbolically.

I've just finished a small thing that actually solves this issue. The existing simplification code looks at equalities with a single free variable. I have added code that looks at all equalities, independent of number of variables, for each equality solves for one at a time and replaces all possible solutions to this variable in all other expressions. Three things may then happen for an expression: 1) All solutions leads to a False result => return False, 2) All solutions leads to a True result => the expression is redundant so can be dropped, 3) None of the above => keep the expression.

Problem is that it relies on `solve` and I believe that in certain cases it can be quite time consuming (it is of course quite easy to replace with `solveset`, but either way, if there are complicated expressions, it will be quite time consuming). I could do some better pruning (like once an expression is deemed redundant it doesn't have to be evaluated again) and probably catch some exceptions from `solve`, although I have not encountered any so far with my limited testing. Also, I have not yet added a test for simplified expressions (Case 3) can be split into "all solutions lead to the same new expression, in that case check if it is simpler than the previous expression" and "keep the expression".)

I happened to build it on top of #21854, so I will open a new PR once that is added (since this additional code may be more doubtful I do not think it is a good idea to add it to the same PR).

It solves this issue and quite often other related things.

**Comment by oscarbenjamin** [comment]
We shouldn't use `solve` for this because it's too slow. Also solving one by one for all symbols in all equations could be very slow if there are many symbols and equations.

Instead I would propose something like:

1. We start with a system of inequalities, unequalities and equations
2. Identify a subset of the equations that are linear
3. Solve the linear equations to eliminate some unknowns
4. Substitute the result into all of the other inequalities and unequalities
5. Identify the linear inequalities
6. Solve the linear inequalities using linear programming
7. Now that all linear equations and inequalities are solved we are left with nonlinear equations/inequalities and also the unequalities - check if any of those are true or false.

Steps 1-4 (and 7) can be done now without an implementation of linear programming and would solve this issue.

Example:
```python
In [14]: i = Symbol("i", integer=True)
    ...: n = Symbol("n", integer=True)
    ...: expr = And(Eq(i, 2 * n), Le(i, 2*n -1))

In [15]: equalities = [Eq(i, 2*n)]

In [16]: inequalities = [Le(i, 2*n-1)]

In [18]: symbols = [i, n]

In [19]: linsolve(equalities, symbols)
Out[19]: {(2⋅n, n)}

In [20]: rep = dict(zip(symbols, linsolve(equalities, symbols).args[0]))

In [21]: rep
Out[21]: {i: 2⋅n, n: n}

In [22]: [ineq.subs(rep) for ineq in inequalities]
Out[22]: [False]
```

**Comment by oscargus** [comment]
As always it depends on the size of the problem how long it will take. The benefit of `solve`, at least in the way I used it, is that it will actually simplify things in a way that is not really possible otherwise. Most constraints resulting from integrals and so on are not really linear. (There are even more ways to deal with multiple solutions I realized, which is some cases can be beneficial as they provide further simplifications, e.g. `And(Eq(x**2, 2), x <= 1)` can be simplified to `Eq(x, -2)` with some further work.)

One thing missing in the pseudo-code is detecting if the expressions are linear or not.  I'm sure it is because it is not needed here, but I encountered that problem in #21869 and it really messed up the results.

Anyway, clearly, running the linear parts first is of course more efficient, but it can indeed make sense to run the non-linear parts as well. Some expressions get quite messy from e.g. `piecewise_folding`, so for me spending a bit of time to obtain a really efficient simplification should be worth it. But easily obtained using some keyword argument.

Btw, I found this paper which seems to provide a quite good overview of different methods to identify redundant linear constraints: https://www.hindawi.com/journals/mpe/2010/723402/ It seems like the following may be a good option: https://www.tandfonline.com/doi/abs/10.1080/00207160601014148 but I cannot get access to it. However, it is briefly described in the first paper. Also, the paper https://dl.acm.org/doi/abs/10.5555/164501.164510 seems relevant. The problem in using linear programming is really that it is not that fast to start with. So a better method for removing redundant constraints would indeed probably be efficient for the linear programming as well. Removing redundant constraints is typically provided as a pre-processing step in most linear programming codes.
